File: backend/app/models/budget.py
"""Modelo Budget: presupuesto mensual por categoria.
Referencia: Solution Design, ERD p.11, DB Schema p.12 — tabla BUDGET.
Constraint UNIQUE(user_id, category_id, month, year): un presupuesto por categoria/mes.
"""

from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Integer, Numeric, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Budget(Base):
    __tablename__ = "budgets"
    __table_args__ = (
        UniqueConstraint("user_id", "category_id", "month", "year", name="uq_budget_per_category_month"),
    )

    budget_id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.user_id"), nullable=False)
    category_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("categories.category_id"), nullable=False
    )
    month: Mapped[int] = mapped_column(Integer, nullable=False)
    year: Mapped[int] = mapped_column(Integer, nullable=False)
    limit_amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, default=func.now()
    )

    user = relationship("User", back_populates="budgets")
    category = relationship("Category", back_populates="budgets")

File: frontend_v2/lib/model/budget.dart
/// Modelo de presupuesto mensual. Ref: Solution Design, DB Schema p.12 — tabla BUDGET.
/// Incluye campos calculados: spent y progress (devueltos por la API).
class Budget {
  final int budgetId;
  final int userId;
  final int categoryId;
  final int month;
  final int year;
  final double limitAmount;
  final DateTime createdAt;
  final String? categoryName;
  final double spent;
  final double progress;

  const Budget({
    required this.budgetId,
    required this.userId,
    required this.categoryId,
    required this.month,
    required this.year,
    required this.limitAmount,
    required this.createdAt,
    this.categoryName,
    this.spent = 0.0,
    this.progress = 0.0,
  });

  factory Budget.fromJson(Map<String, dynamic> json) => Budget(
        budgetId: json['budget_id'] as int,
        userId: json['user_id'] as int,
        categoryId: json['category_id'] as int,
        month: json['month'] as int,
        year: json['year'] as int,
        limitAmount: double.parse(json['limit_amount'].toString()),
        createdAt: DateTime.parse(json['created_at'] as String),
        categoryName: json['category_name'] as String?,
        spent: double.parse((json['spent'] ?? '0.00').toString()),
        progress: (json['progress'] as num?)?.toDouble() ?? 0.0,
      );

  Map<String, dynamic> toJson() => {
        'category_id': categoryId,
        'limit_amount': limitAmount,
        'month': month,
        'year': year,
      };
}
